radix-engine/src/kernel/kernel.rs

```diff
@@ -82,67 +69,35 @@ where
             current_frame: CallFrame::new_root(),
             prev_frame_stack: vec![],
             module,
-        };
+        }
+    }
 
-        // Initial authzone
-        // TODO: Move into module initialization
-        kernel
-            .execute_in_mode::<_, _, RuntimeError>(ExecutionMode::AuthModule, |api| {
-                let auth_zone = AuthZoneStackSubstate::new(
-                    vec![],
-                    auth_zone_params.virtualizable_proofs_resource_addresses,
-                    auth_zone_params.initial_proofs.into_iter().collect(),
-                );
-                let node_id = api.allocate_node_id(RENodeType::AuthZoneStack)?;
-                api.create_node(
-                    node_id,
-                    RENodeInit::AuthZoneStack(auth_zone),
-                    BTreeMap::new(),
-                )?;
-                Ok(())
-            })
-            .expect("AuthModule failed to initialize");
-
-        kernel
-            .execute_in_mode::<_, _, RuntimeError>(ExecutionMode::LoggerModule, |api| {
-                LoggerModule::initialize(api)
-            })
-            .expect("Logger failed to initialize");
-
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Resource(RADIX_TOKEN)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Resource(SYSTEM_TOKEN)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Resource(ECDSA_SECP256K1_TOKEN)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Resource(EDDSA_ED25519_TOKEN)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Resource(PACKAGE_TOKEN)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Component(EPOCH_MANAGER)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Component(CLOCK)),
-            RENodeVisibilityOrigin::Normal,
-        );
-        kernel.current_frame.add_stored_ref(
-            RENodeId::Global(GlobalAddress::Package(FAUCET_PACKAGE)),
-            RENodeVisibilityOrigin::Normal,
-        );
-
-        kernel
+    pub fn initialize(&mut self) -> Result<(), RuntimeError> {
+        self.execute_in_mode::<_, _, RuntimeError>(ExecutionMode::KernelModule, |api| {
+            KernelModuleMixer::on_init(api)
+        })
+    }
+
+    pub fn teardown(mut self) -> (KernelModuleMixer, Option<RuntimeError>) {
+        // Rewind call stack
+        loop {
```

Comment:
Do we need to verify that there's only one frame?

Reply:
It's checked by `if let Some(f) = self.prev_frame_stack.pop()`, unless I missed something.

Reply:
Ah, what I mean is does `teardown` need to enforce the invariant that there's only one frame left on the stack as was done previously?

Reply:
I think it's about the lifecycle of kernel modules, and whether execution should be allowed unpon runtime error.

A typical kernel module, like `auth`/`logger`/`tx_runtime`,  would create some node `on_init` and drop it `on_teardown`.

Here, I've chosen to rewind the call stack and allowed kernel module to teardown.

But looking at my implementation more, the native call stack rewind may not be sufficient to correct the kernel state, so I will be very happy to enforce the invariants 
-  **(A) Only one frame is left on stack** 
-  **(B) No runtime error has ever occurred**.